database/metric_correlations.c

```diff
@@ -0,0 +1,290 @@
+            for (j = 0, d = rb->st->dimensions ; d && j < rb->d ; ++j, d = d->next) {
+                calculated_number *cn = &rb->v[ c * rb->d ];
+                if (!c) {
+                    //TODO use points from query
+                    pd[j].dimension = strdupz (d->name);
+                    pd[j].baseline[c] = cn[j];
+                } else {
+                    pd[j].baseline[c] = cn[j];
+                }
+            }
+        }
+    }
+    rrdr_free(rb);
+    if (!pd)
+        return 0;
+
+    //get highlight
+    RRDR *rh = rrd2rrdr(st, max_points, highlight_after, highlight_before, group_method, group_time, options, NULL, context_param_list);
+    if(!rh) {
+        info("Cannot generate metric correlations output with these parameters on this chart.");
+        freez(pd);
+        return 0;
+    } else {
+        if (rh->d != b_dims) {
+            //TODO handle different dims
+            rrdr_free(rh);
+            freez(pd);
+            return 0;
+        }
+        highlight_points = rrdr_rows(rh);
+        for (c = 0; c != rrdr_rows(rh) ; ++c) {
+            RRDDIM *d;
+            for (j = 0, d = rh->st->dimensions ; d && j < rh->d ; ++j, d = d->next) {
+                calculated_number *cn = &rh->v[ c * rh->d ];
+                    pd[j].highlight[c] = cn[j];
+            }
+        }
+    }
+    rrdr_free(rh);
+
+    for (i = 0; i < b_dims; i++) {
+        fill_nan(&pd[i], highlight_points, baseline_points);
+    }
+
+    for (i = 0; i < b_dims; i++) {
+        run_diffs_and_rev(&pd[i], highlight_points, baseline_points);
+    }
+
+    double d=0, prob=0;
+    for (i=0;i < j ;i++) {
+        if (baseline_points && highlight_points) {
+            kstwo(pd[i].baseline_diffs, baseline_points-1, pd[i].highlight_diffs, highlight_points-1, &d, &prob);
+            buffer_sprintf(wb, "\t\t\t\t\"%s\": %f", pd[i].dimension, prob);
+            if (i != j-1)
+                buffer_sprintf(wb, ",\n");
+            else
+                buffer_sprintf(wb, "\n");
+        }
+    }
+
+    freez(pd);
+    return j;
+}
+
+void metric_correlations (RRDHOST *host, BUFFER *wb, long long baseline_after, long long baseline_before, long long highlight_after, long long highlight_before, long long max_points)
+{
+    info ("Running metric correlations, highlight_after: %lld, highlight_before: %lld, baseline_after: %lld, baseline_before: %lld, max_points: %lld", highlight_after, highlight_before, baseline_after, baseline_before, max_points);
+
+    if (!enable_metric_correlations) {
+        error("Metric correlations functionality is not enabled.");
+        buffer_strcat(wb, "{\"error\": \"Metric correlations functionality is not enabled.\" }");
+        return;
+    }
+
+    if (highlight_before <= highlight_after || baseline_before <= baseline_after) {
+        error("Invalid baseline or highlight ranges.");
+        buffer_strcat(wb, "{\"error\": \"Invalid baseline or highlight ranges.\" }");
+        return;
+    }
```

Comment:
We also need to clamp `baseline_after`, `baseline_before`, `highlight_after`, `highlight_before`, to some `[min, max]` range and (a) either run metric correlation, or (b) return an appropriate error message.

Reply:
Not sure yet what constitutes invalid range, aside from obvious being the before before after and such. Perhaps we will add, e.g. if we don't have any metrics during that period? But I guess since the input will come from the cloud which will be highlighted on a region containing metrics, it should be ok. We will see, I'm expecting the cloud to kinda safe-guard the values, but we will protect if we want, yes.

Reply:
What I mean is that we should work with @andrewm4894, in order to come up with the largest duration we are willing to allow the agents to serve. IOW, we don't want to run metric correlation for a duration of 1 month (as this will most probably break the agent in some way).

> `error("Invalid baseline or highlight ranges.");`

This is a coalesced error message. It's not clear which invalid case we actually hit from the message description. We should return a separate one for an invalid `baseline` range and another one for the `highlight` range.

> We will see, I'm expecting the cloud to kinda safe-guard the values, but we will protect if we want, yes

Agreed, but if we don't have a predefined limit someone could DoS the agent.

Reply:
What I mean is that we should work with @andrewm4894, in order to come up with the largest duration we are willing to allow the agents to serve. IOW, we don't want to run metric correlation for a duration of 1 month (as this will most probably break the agent in some way).

> `error("Invalid baseline or highlight ranges.");`

This is a coalesced error message. It's not clear which invalid case we actually hit from the message description. We should return a separate one for an invalid `baseline` range and another one for the `highlight` range.

> We will see, I'm expecting the cloud to kinda safe-guard the values, but we will protect if we want, yes

Agreed, but if we don't have a predefined limit someone could DoS the agent.

Reply:
Indeed. That part needs attention, and will be done right after. As is with the max_points, we will define a good number to balance performance and accuracy, but that will come I guess from testing it in real world.